contracts/protocol/tokenization/libraries/MintableERC721Logic.sol

```diff
@@ -399,18 +494,99 @@ library MintableERC721Logic {
+            uint256 oldMultiplier = erc721Data.traitsMultipliers[tokenIds[i]];
+            erc721Data.traitsMultipliers[tokenIds[i]] = multipliers[i];
+            if (owner == address(0)) {
+                continue;
+            }
+
+            bool isAtomicPrev = oldMultiplier != 0 &&
+                oldMultiplier != WadRayMath.WAD;
+            bool isAtomicNext = multipliers[i] != WadRayMath.WAD;
+
+            if (isAtomicPrev && !isAtomicNext) {
+                _removeTokenFromOwnerEnumeration(
+                    erc721Data,
+                    owner,
+                    tokenIds[i],
+                    erc721Data.userState[owner].atomicBalance,
+                    isAtomicPrev
+                );
+                _addTokenToOwnerEnumeration(
+                    erc721Data,
+                    owner,
+                    tokenIds[i],
+                    erc721Data.userState[owner].balance,
+                    isAtomicNext
+                );
+
+                erc721Data.userState[owner].atomicBalance -= 1;
+                erc721Data.userState[owner].balance += 1;
+
+                if (erc721Data.isUsedAsCollateral[tokenIds[i]]) {
+                    erc721Data
+                        .userState[owner]
+                        .atomicCollateralizedBalance -= 1;
+                    erc721Data.userState[owner].collateralizedBalance += 1;
+                }
+            } else if (!isAtomicPrev && isAtomicNext) {
+                _removeTokenFromOwnerEnumeration(
+                    erc721Data,
+                    owner,
+                    tokenIds[i],
+                    erc721Data.userState[owner].balance,
+                    isAtomicPrev
+                );
+                _addTokenToOwnerEnumeration(
+                    erc721Data,
+                    owner,
+                    tokenIds[i],
+                    erc721Data.userState[owner].atomicBalance,
+                    isAtomicNext
+                );
+
+                erc721Data.userState[owner].balance -= 1;
+                erc721Data.userState[owner].atomicBalance += 1;
+
+                if (erc721Data.isUsedAsCollateral[tokenIds[i]]) {
+                    erc721Data.userState[owner].collateralizedBalance -= 1;
+                    erc721Data
+                        .userState[owner]
+                        .atomicCollateralizedBalance += 1;
+                }
+            }
+        }
+    }
+
     function _checkBalanceLimit(
         MintableERC721Data storage erc721Data,
-        bool ATOMIC_PRICING,
         uint64 balance
     ) private view {
-        if (ATOMIC_PRICING) {
-            uint64 balanceLimit = erc721Data.balanceLimit;
-            require(
-                balanceLimit == 0 || balance <= balanceLimit,
-                Errors.NTOKEN_BALANCE_EXCEEDED
-            );
-        }
+        uint64 balanceLimit = erc721Data.balanceLimit;
+        require(
+            balanceLimit == 0 || balance <= balanceLimit,
```

Comment:
better to limit only atomic portion, but UniV3 is an exception because it's stored in ownedTokens but it has atomic pricing, so we need two terms:

- trait boosted
- uni v3